specification/postgresql/resource-manager/Microsoft.DBforPostgreSQL/preview/2023-06-01-preview/Migrations.json

```diff
@@ -0,0 +1,1254 @@
+      ],
+      "type": "string",
+      "description": "Migration details level.",
+      "x-ms-enum": {
+        "name": "MigrationDetailsLevel",
+        "modelAsString": true
+      }
+    },
+    "MigrationMode": {
+      "enum": [
+        "Offline",
+        "Online"
+      ],
+      "type": "string",
+      "description": "There are two types of migration modes Online and Offline",
+      "x-ms-enum": {
+        "name": "MigrationMode",
+        "modelAsString": true
+      }
+    },
+    "MigrationOption": {
+      "enum": [
+        "Validate",
+        "Migrate",
+        "ValidateAndMigrate"
+      ],
+      "type": "string",
+      "description": "Supported types of migration request include Validate, Migrate and ValidateAndMigrate",
+      "x-ms-enum": {
+        "name": "MigrationOption",
+        "modelAsString": true
+      }
+    },
+    "MigrationState": {
+      "enum": [
+        "InProgress",
+        "WaitingForUserAction",
+        "Canceled",
+        "Failed",
+        "Succeeded",
+        "ValidationFailed",
+        "CleaningUp"
+      ],
+      "type": "string",
+      "description": "Migration state.",
+      "x-ms-enum": {
+        "name": "MigrationState",
+        "modelAsString": true
+      }
+    },
+    "MigrationStatus": {
+      "type": "object",
+      "properties": {
+        "state": {
+          "$ref": "#/definitions/MigrationState",
+          "readOnly": true,
+          "description": "State of migration"
+        },
+        "error": {
+          "type": "string",
+          "readOnly": true,
+          "description": "Error message, if any, for the migration state"
+        },
+        "currentSubStateDetails": {
+          "$ref": "#/definitions/MigrationSubStateDetails",
+          "readOnly": true,
+          "description": "Current Migration sub state details."
+        }
+      },
+      "description": "Migration status."
+    },
+    "MigrationSubState": {
+      "enum": [
+        "PerformingPreRequisiteSteps",
+        "WaitingForLogicalReplicationSetupRequestOnSourceDB",
+        "WaitingForDBsToMigrateSpecification",
+        "WaitingForTargetDBOverwriteConfirmation",
+        "WaitingForDataMigrationScheduling",
+        "WaitingForDataMigrationWindow",
+        "MigratingData",
+        "WaitingForCutoverTrigger",
+        "CompletingMigration",
+        "Completed",
+        "CancelingRequestedDBMigrations",
+        "ValidationInProgress"
+      ],
+      "type": "string",
+      "description": "Migration sub state.",
+      "x-ms-enum": {
+        "name": "MigrationSubState",
+        "modelAsString": true
+      }
+    },
+    "MigrationSubStateDetails": {
+      "type": "object",
+      "properties": {
+        "currentSubState": {
+          "$ref": "#/definitions/MigrationSubState",
+          "readOnly": true
+        },
+        "dbDetails": {
+          "type": "object",
+          "additionalProperties": {
+            "$ref": "#/definitions/DbMigrationStatus"
```

Comment:
use of additionalProperties is not allowed for properties owned by the service. The only time its ok to use it is when the properties are pass thru (user defined) and not subject to any validations. Please make this an array or provide an explanation for why you need this.

Reply:
DbDetails is a dictionary at server side. Hence had to use additional properties here as we need to provide DbMigrationStatus object against each databaseName as a key. Hope this answers your question.

Reply:
Please find below the sample response, where DbDetails object is highlighted in BOLD

```
{
	"properties": {
		"migrationId": "11d99ce5-84d5-4adf-82a0-cbde07df3a5a",
		"currentStatus": {
			"state": "Succeeded",
			"currentSubStateDetails": {
				"currentSubState": "Completed",
				**"dbDetails": {
					"newdb": {
						"databaseName": "newdb",
						"migrationState": "Success",
						"startedOn": "2023-08-08T07:58:29.5349158+00:00",
						"endedOn": "2023-08-08T07:58:37.4522655+00:00",
						"fullLoadErroredTables": 0,
						"fullLoadQueuedTables": 0,
						"fullLoadLoadingTables": 0,
						"fullLoadCompletedTables": 0,
						"cdcUpdateCounter": 0,
						"cdcDeleteCounter": 0,
						"cdcInsertCounter": 0,
						"appliedChanges": 0,
						"incomingChanges": 0,
						"latency": 0
					}
				},**
				"validationDetails": {
					"status": "Warning",
					"validationStartTimeInUtc": "2023-08-08T02:27:17.8238008Z",
					"validationEndTimeInUtc": "2023-08-08T02:27:59.4987788Z",
					"serverLevelValidationDetails": [
						{
							"type": "AuthenticationAndConnectivityValidation",
							"state": "Succeeded"
						},
						{
							"type": "SkuValidation",
							"state": "Warning",
							"messages": [
								{
									"state": "Warning",
									"message": "The storage provisioned for target server is less than the provisioned storage of the source server. Consider provisioning a storage higher than your source for the target server and run the migration."
								}
							]
						}
					],
					"dbLevelValidationDetails": [
						{
							"databaseName": "newdb",
							"startedOn": "2023-08-08T07:57:20.5481619+00:00",
							"endedOn": "2023-08-08T07:57:20.9557345+00:00",
							"summary": [
								{
									"type": "ExtensionsValidation",
									"state": "Succeeded"
								}
							]
						}
					]
				}
			}
		},
		"migrationMode": "Offline",
		"migrationWindowEndTimeInUtc": "2023-08-08T08:00:33.2347243Z",
		"sourceDbServerResourceId": "/subscriptions/ffffffff-ffff-ffff-ffff-ffffffffffff/resourceGroups/pmv-test-rg/providers/Microsoft.DBforPostgreSQL/servers/pmv-sspg-test",
		"targetDbServerResourceId": "/subscriptions/929287ae-832a-4946-8006-a6cc2a3f7244/resourceGroups/spkota-oneboxrg301/providers/Microsoft.DBforPostgreSQL/flexibleServers/pkota-oneboxserve202",
		"dbsToMigrate": [
			"newdb"
		],
		"setupLogicalReplicationOnSourceDbIfNeeded": "True",
		"overwriteDbsInTarget": "True",
		"migrationWindowStartTimeInUtc": "2023-08-08T07:55:56.7837686Z",
		"triggerCutover": "True",
		"sourceDbServerMetadata": {
			"location": "southcentralus",
			"storageMb": 131072,
			"sku": {}
		},
		"targetDbServerMetadata": {
			"location": "Onebox",
			"storageMb": 65536,
			"sku": {
				"name": "Standard_D8s_v3",
				"tier": "Standard_D8s_v3"
			}
		},
		"migrationOption": "ValidateAndMigrate",
                "sourceType": "PostgreSQLSingleServer",
                "sslMode": "VerifyFull"
	},
	"location": "Onebox",
	"id": "/subscriptions/929287ae-832a-4946-8006-a6cc2a3f7244/resourceGroups/spkota-oneboxrg301/providers/Microsoft.DBforPostgreSQL/flexibleServers/pkota-oneboxserve202/migrations/213",
	"name": "213",
	"type": "Microsoft.DBforPostgreSQL/flexibleServers/migrations"
}
```

Reply:
As discussed this API implementation is in use by many customers and unfortunately this issue was not caught earlier. Hence marking this as resolved.

Reply:
Going forward this pattern will no longer be allowed. Approving this time as there are existing implementations following the same pattern.